File: data_structures/stacks_and_queues/queue.py
class Node():
  def __init__(self, value):
    self.value = value
    self.next = None

# Queue - Linked List implementation
class Queue():
  def __init__(self):
    self.first = None
    self.last = None
    self.length = 0

  def __str__(self):
    return str(self.__dict__)

  def printlist(self):
    myarray = []
    currentNode = self.first
    while currentNode:
      myarray.append(currentNode.value)
      currentNode = currentNode.next
    return myarray

  def peek(self):
    return self.first.value

  def enqueue(self, value):
    new_node = Node(value)
    if self.length == 0:
      self.first = new_node
      self.last = new_node
    else:
      self.last.next = new_node
      self.last = new_node
    self.length += 1

  def dequeue(self):
    if not self.first:
      return None
    if self.first == self.last:
      self.last = None
    self.first = self.first.next
    self.length -= 1

queue = Queue()
queue.enqueue("David")
print(queue.printlist())
queue.enqueue("Jimmy")
print(queue.printlist())
queue.enqueue("Mary")
print(queue.printlist())
queue.dequeue()
print(queue.printlist())
print(queue.peek())

File: data_structures/stacks_and_queues/Queue.java
package data_structures.stacks_and_queues;

class Node {
    String value;
    Node next;
    Node(String value) {
        this.value = value;
        this.next = null;
    }
}

class Queue {
    private Node first;
    private Node last;
    private int length;

    Queue() {
        this.first = null;
        this.last = null;
        this.length = 0;
    }

    public String peek() {
        return first != null ? first.value : null;
    }

    public void enqueue(String value) {
        Node newNode = new Node(value);
        if (length == 0) {
            first = newNode;
            last = newNode;
        } else {
            last.next = newNode;
            last = newNode;
        }
        length++;
    }

    public void dequeue() {
        if (first == null) {
            return;
        }
        if (first == last) {
            last = null;
        }
        first = first.next;
        length--;
    }

    public String[] printList() {
        String[] result = new String[length];
        int index = 0;
        Node currentNode = first;
        while (currentNode != null) {
            result[index++] = currentNode.value;
            currentNode = currentNode.next;
        }
        return result;
    }

    // Returns a string representation of the queue for debugging purposes
    @Override
    public String toString() {
        StringBuilder result = new StringBuilder("{");
        result.append("\"first\": ").append(first != null ? first.value : "null").append(", ");
        result.append("\"last\": ").append(last != null ? last.value : "null").append(", ");
        result.append("\"length\": ").append(length);
        result.append("}");
        return result.toString();
    }

    public static void main(String[] args) {
        Queue queue = new Queue();
        queue.enqueue("David");
        System.out.println(java.util.Arrays.toString(queue.printList()));
        queue.enqueue("Jimmy");
        System.out.println(java.util.Arrays.toString(queue.printList()));
        queue.enqueue("Mary");
        System.out.println(java.util.Arrays.toString(queue.printList()));
        queue.dequeue();
        System.out.println(java.util.Arrays.toString(queue.printList()));
        System.out.println("Peek: " + queue.peek());
        System.out.println("toString: " + queue.toString());
    }
}
